### PR TITLE
Use C++20

### PR DESCRIPTION
**Pull request by SiarheiFedartsou**
I believe it is a time to bump C++ standard in this project. It has quite a lot of useful things OSRM can benefit from(e.g. std::span and also I see that even `json_render` benchmark became better due to some reason :) ). I hope there are no objections for this.

## Benchmark Results
| Benchmark | Base | PR |
|-----------|------|----|
| alias | aliased u32: 1094.05<br/>plain u32: 1092.11<br/>aliased double: 960.248<br/>plain double: 947.363 | aliased u32: 1091.75<br/>plain u32: 1089.74<br/>aliased double: 959.988<br/>plain double: 948.72 |
| json-render | String: 8.2719ms<br/>Stringstream: 11.5894ms<br/>Vector: 7.47187ms | String: 6.75741ms<br/>Stringstream: 10.5255ms<br/>Vector: 6.62243ms |
| match_ch | Default radius: <br/>4.43728ms/req at 82 coordinate<br/>0.0541131ms/coordinate<br/>Radius 5m: <br/>4.43271ms/req at 82 coordinate<br/>0.0540574ms/coordinate<br/>Radius 10m: <br/>15.0005ms/req at 82 coordinate<br/>0.182933ms/coordinate<br/>Radius 15m: <br/>36.5988ms/req at 82 coordinate<br/>0.446327ms/coordinate<br/>Radius 30m: <br/>312.388ms/req at 82 coordinate<br/>3.80961ms/coordinate | Default radius: <br/>4.53578ms/req at 82 coordinate<br/>0.0553144ms/coordinate<br/>Radius 5m: <br/>4.53172ms/req at 82 coordinate<br/>0.0552649ms/coordinate<br/>Radius 10m: <br/>15.343ms/req at 82 coordinate<br/>0.18711ms/coordinate<br/>Radius 15m: <br/>37.2938ms/req at 82 coordinate<br/>0.454803ms/coordinate<br/>Radius 30m: <br/>319.071ms/req at 82 coordinate<br/>3.89112ms/coordinate |
| match_mld | Default radius: <br/>2.88705ms/req at 82 coordinate<br/>0.035208ms/coordinate<br/>Radius 5m: <br/>2.80234ms/req at 82 coordinate<br/>0.0341749ms/coordinate<br/>Radius 10m: <br/>10.1313ms/req at 82 coordinate<br/>0.123553ms/coordinate<br/>Radius 15m: <br/>25.6745ms/req at 82 coordinate<br/>0.313104ms/coordinate<br/>Radius 30m: <br/>297.859ms/req at 82 coordinate<br/>3.63243ms/coordinate | Default radius: <br/>2.87119ms/req at 82 coordinate<br/>0.0350145ms/coordinate<br/>Radius 5m: <br/>2.86241ms/req at 82 coordinate<br/>0.0349074ms/coordinate<br/>Radius 10m: <br/>10.4629ms/req at 82 coordinate<br/>0.127596ms/coordinate<br/>Radius 15m: <br/>26.7566ms/req at 82 coordinate<br/>0.3263ms/coordinate<br/>Radius 30m: <br/>314.468ms/req at 82 coordinate<br/>3.83497ms/coordinate |
| packedvector | random write:<br/>std::vector 9813.3 ms<br/>util::packed_vector 74360.7 ms<br/>slowdown: 7.57755<br/>random read:<br/>std::vector 8675.09 ms<br/>util::packed_vector 30857.7 ms<br/>slowdown: 3.55704 | random write:<br/>std::vector 9754.69 ms<br/>util::packed_vector 81460.1 ms<br/>slowdown: 8.35087<br/>random read:<br/>std::vector 8478.48 ms<br/>util::packed_vector 31332 ms<br/>slowdown: 3.69548 |
| route_ch | 1000 routes, 3 coordinates, no alternatives, overview=full, steps=true<br/>597.456ms<br/>0.597456ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=full, steps=true<br/>379.2ms<br/>0.3792ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=full, steps=true<br/>763.511ms<br/>0.763511ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false<br/>161.229ms<br/>0.161229ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false<br/>103.889ms<br/>0.103889ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false<br/>146.011ms<br/>0.146011ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>161.279ms<br/>0.161279ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>103.87ms<br/>0.10387ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false, radius=750<br/>144.998ms<br/>0.144998ms/req | 1000 routes, 3 coordinates, no alternatives, overview=full, steps=true<br/>596.23ms<br/>0.59623ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=full, steps=true<br/>376.829ms<br/>0.376829ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=full, steps=true<br/>763.386ms<br/>0.763386ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false<br/>161.562ms<br/>0.161562ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false<br/>104.426ms<br/>0.104426ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false<br/>144.873ms<br/>0.144873ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>160.242ms<br/>0.160242ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>104.019ms<br/>0.104019ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false, radius=750<br/>144.505ms<br/>0.144505ms/req |
| route_mld | 1000 routes, 3 coordinates, no alternatives, overview=full, steps=true<br/>722.484ms<br/>0.722484ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=full, steps=true<br/>467.443ms<br/>0.467443ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=full, steps=true<br/>946.061ms<br/>0.946061ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false<br/>280.552ms<br/>0.280552ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false<br/>168.964ms<br/>0.168964ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false<br/>304.268ms<br/>0.304268ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>280.978ms<br/>0.280978ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>168.926ms<br/>0.168926ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false, radius=750<br/>300.273ms<br/>0.300273ms/req | 1000 routes, 3 coordinates, no alternatives, overview=full, steps=true<br/>727.123ms<br/>0.727123ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=full, steps=true<br/>469.123ms<br/>0.469123ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=full, steps=true<br/>956.257ms<br/>0.956257ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false<br/>286.806ms<br/>0.286806ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false<br/>167.625ms<br/>0.167625ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false<br/>306.994ms<br/>0.306994ms/req<br/>1000 routes, 3 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>280.106ms<br/>0.280106ms/req<br/>1000 routes, 2 coordinates, no alternatives, overview=false, steps=false, radius=750<br/>167.448ms<br/>0.167448ms/req<br/>1000 routes, 2 coordinates, 3 alternatives, overview=false, steps=false, radius=750<br/>305.274ms<br/>0.305274ms/req |
| rtree | 1 result:<br/>208.935ms ->  0.0208935 ms/query<br/>10 results:<br/>244.479ms ->  0.0244479 ms/query | 1 result:<br/>207.379ms ->  0.0207379 ms/query<br/>10 results:<br/>242.495ms ->  0.0242495 ms/query |

### CODE REVIEW

File: .github/workflows/osrm-backend.yml [hunk 1]
```diff
@@ -396,7 +396,36 @@ jobs:
...
+          /usr/local/include/boost
+          /usr/local/lib/libboost*
+        key: v1-boost-${{ runner.os }}-${{ runner.arch }}-${{ matrix.runs-on }}
+        restore-keys: |
+          v1-boost-${{ runner.os }}-${{ runner.arch }}-${{ matrix.runs-on }}
 
+    - name: Install Boost
+      if: steps.cache-boost.outputs.cache-hit != 'true' && runner.os == 'Linux' && matrix.ENABLE_CONAN != 'ON'
+      run: |
+        BOOST_VERSION="1.85.0"
```
Comment (SiarheiFedartsou) [inline .github/workflows/osrm-backend.yml:419]: There were problems with Boost from apt on Ubuntu 22.04, so I propose to install it manually for the time being.

File: .clang-tidy [hunk 2]
```diff
@@ -82,6 +83,7 @@ Checks: >
     -readability-make-member-function-const,
     -readability-redundant-string-init,
     -readability-non-const-parameter,
+    -readability-container-contains,
```
Comment (DennisOSRM) [inline .clang-tidy:86]: How about adding a ticket into the backlog to fix these warnings?

File: .github/workflows/osrm-backend.yml (hunk 1 above)
Comment (DennisOSRM) [reply .github/workflows/osrm-backend.yml:419]: Would a switch to 24.04 make this easier?

**Review by DennisOSRM — APPROVED** [review]

File: .clang-tidy (hunk 2 above)
Comment (SiarheiFedartsou) [reply .clang-tidy:86]: Sure, https://github.com/Project-OSRM/osrm-backend/issues/6902

**Comment by DennisOSRM** [comment]
Great change!

File: .github/workflows/osrm-backend.yml (hunk 1 above)
Comment (SiarheiFedartsou) [reply .github/workflows/osrm-backend.yml:419]: Well, with a switch to 24.04 I had another challenge: sol2 is not compiling with [newest clang we would have to switch to on 24.04](https://github.com/ThePhD/sol2/issues/1581). There is already fix in LLVM, so I believe it is just easier to wait for a while until this fix will come to GitHub Actions runner image for 24.04 and switch.

Comment (cuavas) [reply .github/workflows/osrm-backend.yml:419]: The fix is only going to make it to clang 19, unfortunately.

**Pull request merged**